river/drift/adwin_c.pyx

```diff
@@ -0,0 +1,379 @@
+                    self.bucket_deque.append(Bucket())
+                    next_bucket = self.bucket_deque[-1]
+                n1 = self._bucket_size(idx)
+                n2 = self._bucket_size(idx)
+                u1 = bucket.get_total_at(0) / n1
+                u2 = bucket.get_total_at(1) / n2
+                incremental_variance = n1 * n2 * ((u1 - u2) * (u1 - u2)) / (n1 + n2)
+                next_bucket.insert_data(
+                    bucket.get_total_at(0) + bucket.get_total_at(1),
+                    bucket.get_variance_at(1) + incremental_variance
+                )
+                self.n_buckets += 1
+                bucket.compress(2)
+
+                if next_bucket.current_idx <= self.MAX_BUCKETS:
+                    break
+            else:
+                break
+
+
+    cdef bint _detect_change(self):
+        """Detect concept change.
+
+        This function is responsible for analysing different cutting points in
+        the sliding window, to verify if there is a significant change.
+
+        Returns
+        -------
+        bint
+            If True then a change is detected.
+
+        Notes
+        -----
+        If change was detected, one should verify the new window size,
+        by reading the width property.
+
+        """
+        cdef:
+            int idx, k
+            bint change_detected, exit_flag
+            double n0, n1, n2, u0, u1, u2, v0, v1
+            Bucket bucket
+
+        change_detected = False
+        exit_flag = False
+        self.tick += 1
+        n0 = 0
+        if (self.tick % self.clock == 0) and (self.width > self.min_window_len):
+            reduce_width = True
+            while reduce_width:
+                reduce_width = False
+                exit_flag = False
+                n0 = 0.0
+                n1 = self.width
+                u0 = 0.0
+                u1 = self.total
+                v0 = 0
+                v1 = self.variance
+                n2 = 0.0
+                u2 = 0.0
+
+                for idx in range(len(self.bucket_deque) - 1, -1 , -1):
+                    if exit_flag:
+                        break
+                    bucket = self.bucket_deque[idx]
+
+                    for k in range(bucket.current_idx - 1):
+                        n2 = self._bucket_size(idx)
+                        u2 = bucket.get_total_at(k)
+
+                        if n0 > 0.0:
+                            v0 += (
+                                    bucket.get_variance_at(k) + n0 * n2
+                                    * (u0 / n0 - u2 / n2) * (u0 / n0 - u2 / n2)
+                                    / (n0 + n2)
+                            )
+
+                        if n1 > 0.0:
+                            v1 -= (
+                                    bucket.get_variance_at(k) + n1 * n2
+                                    * (u1 / n1 - u2 / n2) * (u1 / n1 - u2 / n2)
+                                    / (n1 + n2)
+                            )
```

Comment:
The same approximation is applied here. The formulae derived from the work of Chan et al. (please check `stats.Var`) produces exact values. You can verify that [here]().

The updates here do not use any second order term (`== (n - ddof) * variance`).

Reply:
We agreed to keep the code as is since `stats.Var` is Python code and is not worth the overhead.